### PR TITLE
virt-operator: [BUGFIX] Retain Kubernetes revision annotation

### PR DESCRIPTION
**Pull request by Barakmor1**
Currently, virt-operator constantly removes the Kubernetes revision
annotation as it is not part of the operator definition in the
installStrategy.
    
This behavior is incorrect because the Kubernetes control plane keeps
adding the revision annotation back, as it is a system annotation that
should not be handled by virt-operator.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-operator/resource/apply/apps_test.go [hunk 1]
```diff
@@ -1166,4 +1168,106 @@ var _ = Describe("Apply Apps", func() {
...
+					return false, nil, err
+				}
+
+				var newDeployment *appsv1.Deployment
+				if err = json.Unmarshal(newDeploymentBytes, &newDeployment); err != nil {
+					return false, nil, err
+				}
+
+				return true, newDeployment, nil
+			})
```
Comment (fossedihelm) [inline pkg/virt-operator/resource/apply/apps_test.go:1256]: You can avoid customizing the behavior and use the fake client directly (it already does this job for you).
It is enough to register the deployment in the fake client (through `Create`) and then check the resource after the `syncDeployment` function.
This is just an example for inspiration: https://github.com/fossedihelm/kubevirt/commit/b5632fd751e1833e76b44104e0a801b0917a4016

N.B: There are couple of switch betweem var/const and drop of unneeded global variables

**Review by fossedihelm — COMMENTED** [review]
Thank you @Barakmor1 for fixing this hard to discover bug.
I have some suggestion below.

File: pkg/virt-operator/resource/apply/apps.go [hunk 2]
```diff
@@ -103,6 +103,16 @@ func (r *Reconciler) syncDeployment(origDeployment *appsv1.Deployment) (*appsv1.
...
+	if existingCopy.ObjectMeta.Annotations != nil {
+		revisionAnnotation := "deployment.kubernetes.io/revision"
+		if val, ok := existingCopy.ObjectMeta.Annotations[revisionAnnotation]; ok {
+			if deployment.ObjectMeta.Annotations == nil {
+				deployment.ObjectMeta.Annotations = map[string]string{}
+			}
+			deployment.ObjectMeta.Annotations[revisionAnnotation] = val
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (Barakmor1) [reply pkg/virt-operator/resource/apply/apps.go]: Always nice to learn new things :)

Thanks!

File: pkg/virt-operator/resource/apply/apps_test.go (hunk 1 above)
Comment (Barakmor1) [reply pkg/virt-operator/resource/apply/apps_test.go]: Thanks for the note i modified the test please have another loop :)

**Review by fossedihelm — APPROVED** [review]
Thank you!

File: pkg/virt-operator/resource/apply/apps.go [hunk 3]
```diff
@@ -103,6 +103,14 @@ func (r *Reconciler) syncDeployment(origDeployment *appsv1.Deployment) (*appsv1.
 		return nil, err
 	}
 
+	revisionAnnotation := "deployment.kubernetes.io/revision"
```
Comment (jean-edouard) [inline pkg/virt-operator/resource/apply/apps.go:106]: This should be a const

**Review by jean-edouard — COMMENTED** [review]
/approve
/hold

File: pkg/virt-operator/resource/apply/apps.go (hunk 3 above)
Comment (Barakmor1) [reply pkg/virt-operator/resource/apply/apps.go]: Done,
Thanks!

**Comment by fossedihelm** [comment]
/lgtm
Thanks

**Comment by Barakmor1** [comment]
/unhold

@jean-edouard Please feel free to add the hold label back if there is something else on your mind.

**Pull request merged**